File: 120/triangle.py
class Solution:
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        n = len(triangle)
        tmp = []

        for i in range(n):
            row = triangle[i]
            for j in range(len(row)):
                row[j] = 0
            tmp.append(row)
            
        last_row = []
        for j in range(n+1):
            last_row.append(0)
        tmp.append(last_row)

        for i in range(n-1, -1, -1):
            for j in range(len(triangle[i])):
                tmp[i][j] = min(tmp[i+1][j], tmp[i+1][j+1]) + triangle[i][j]

        return tmp[0][0]

File: 120/Triangle.scala
object Triangle {
  def minimumTotal(triangle: List[List[Int]]): Int = {
    val n = triangle.length + 1
//    var dp = scala.collection.mutable.ArrayBuffer.fill(n,n)(0)
    val dp = Array.ofDim[Int](n, n)
    for (i <- 1 until n) dp(n-1)(i) = 0
    for (i <- n-1 to 0) {
      for (j <- triangle(i).indices) {
        dp(i)(j) = triangle(i)(j) + (dp(i+1)(j) min dp(i+1)(j))
      }
    }
    dp(0)(0)
  }

  def main(args: Array[String]): Unit = {
    val triangle = List(
      List(2),
      List(3, 4),
      List(6, 5, 7),
      List(4, 1, 8, 3)
    )
    println(minimumTotal(triangle))
  }
}
